### DISCUSSION TITLE
Environment variables confusion

### DISCUSSION THREAD

**Discussion by gwalter98** [vercel/community#1924] [Help] [answered]
### Summary

I am following [this](https://vercel.com/guides/using-vercel-cli-for-custom-workflows) guide about using the Vercel CLI for custom workflows. I'm hoping to set up a build pipeline for my Next.js project in Azure Pipelines. Before this I'm attempting to test the CLI commands on my local machine. I've completed steps 1-5 in the guide without issue so I am logged into the CLI, have created my project, and have saved the specified environment variables to my machine.

However, I'm running into an error when I run `vercel build` but it doesn't happen all the time and I think I've narrowed it down to being a problem with my project's environment variables. I only seem to get a successful build if I leave the `.env` file in my project and don't specify an environment when running `vercel pull`.

### Additional information

I currently have a `.env` file in my Next.js project. I don't have any variables I would consider 'secret' so I have this as a tracked file in my repository. I have also duplicated these variables and stored them in the Vercel Environment Variables UI where I am using the same values (for now) across the Production, Preview, and Deployment environments. 

When I run `vercel pull --environment=preview` and `vercel build` (on my local machine and in the pipeline) I get the following error:
```
info  - Collecting page data ..node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

TypeError [ERR_INVALID_URL]: Invalid URL
    at new NodeError (node:internal/errors:399:5)
    at URL.onParseError (node:internal/url:566:9)
    at new URL (node:internal/url:646:5)
    at parseUrl (C:\repos\website\node_modules\next-auth\utils\parse-url.js:17:16)
    at Object.<anonymous> (C:\repos\website\node_modules\next-auth\react\index.js:72:34)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19) {
  type: 'TypeError',
  input: '',
  code: 'ERR_INVALID_URL'
}

Node.js v18.15.0
Error: Command "npm run build" exited with 1 
```
When running `--debug` I can see: ` Loaded environment variables from "C:\repos\website\.vercel\.env.preview.local"`

But when I run (on my local machine) just `vercel pull` and `vercel build` I get a successful build. But, I can see: `Failed loading environment variables: Error: ENOENT: no such file or directory, open 'C:\repos\website\.vercel\.env.preview.local'`.

If I delete `.env` and run the second set of commands again it also fails to build. 

I'm very confused about what the correct way of managing the environment variables with my Next.js project and Vercel is. I thought that if I don't have a `.env` file in my project running `vercel pull` with the desired environment specified should handle creating the correct environment variables in each case, but that doesn't seem to be happening for me? Can anyone please help me figure out what I'm doing wrong?

I'd greatly appreciate any help with getting a successful build using the Preview environment variables.

### Example

https://gist.github.com/GWALTER4/71585ae7115b22912496a117c16314c7

**Comment by amyegan** [comment, marked as answer]
Hi @GWALTER4 👋 Based on your explanation and error message, I suspect the first error was caused by a typo in one of your environment variables. The first thing to check is that all of your environment variables are present and correct in the Environment Variables UI. Because the error refers to `next-auth`, you should pay extra attention to any related to `next-auth` configuration. 

After double checking your environment variables, be sure to run `vercel pull --environment=preview` and check that the `.vercel/.env.preview.local` file was created/updated with the expected values. After that, I expect `vercel build` should succeed.

With regard to the `vercel pull` behavior you observed, I think this is normal. When you run `vercel pull` without the `environment=preview` flag, it pulls in the Development environment variables as `.vercel/.env.development.local`. 

Since `vercel build` attempts a Preview build by default, it will look for a `.vercel/.env.preview.local`, which doesn't exist unless you have previously run `vercel pull --environment=preview` to generate the file. In that case, you see the warning that `.env.preview.local` could not be found. When running the project locally for development, you should use the `vercel dev` command instead

**Reply by gwalter98** [reply]
Thank you for pointing the error was coming from `next-auth`.  That led me to find the root cause of the issue! Its always the small easy to miss things that cause these kind of problems!

The issue was there was a missing environment variable required by the project (`NEXTAUTH_URL`) which had not been set and required a value. I've only recently inherited the project and didn't realise we had code implemented that required this as it seems to be used internally by a dependency - not directly in the project's codebase.

Now that I've added that environment variable with a value, I'm getting a successful build! 🎉 

What I do find curious though is that before if I was to run just `vercel` in the project's directory, it would successfully build and deploy the project to Vercel. I wonder why this process didn't seem to be affected by the missing environment variable?

**Reply by amyegan** [reply]
When building locally, Next.js had access to [use your local environment variables](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables). If there was a `.env` with a `NEXTAUTH_URL` value in your local repo, then Next.js was able to use that to complete the build successfully
